### PR TITLE
s/mt_rand/random_int/

### LINKED ISSUE #26: Change MT_RAND
Change mt_rand and mt_getrandmax ... etc to cryptographically secure functions 

https://github.com/paragonie/random_compat

Future proofs as well by using PHP 7 functions.

PR will come once the reference project gets tagged 1.0.0

### PR DESCRIPTION
**Pull request by schnittstabil**
Fixes #26.

### CODE REVIEW

File: src/Guard.php [hunk 1]
```diff
@@ -231,38 +231,7 @@ protected function validateToken($name, $value)
      */
     protected function createToken()
     {
-        $token = "";
-
-        if (function_exists("random_bytes")) {
-            $rawToken = random_bytes($this->strength);
-            if ($rawToken !== false) {
```
Comment (schnittstabil) [inline src/Guard.php]: `$rawToken == false` [can not happen](http://php.net/manual/en/function.random-bytes.php#refsect1-function.random-bytes-returnvalues), obviously a copy-paste-mistake of the [`openssl_random_pseudo_bytes` branch](http://php.net/manual/de/function.openssl-random-pseudo-bytes.php#refsect1-function.openssl-random-pseudo-bytes-returnvalues).

**Comment by AndrewCarterUK** [comment]
Just noticed this PR (I've just made a similar one that I'm now closing: https://github.com/slimphp/Slim-Csrf/pull/42).

There are some slight differences between the two which I'll highlight:

1. I've removed the reference to OWASP in the phpdoc. The whole article is awful security practices everywhere and with that bit stripped out there's nothing of theirs that remains.
2. I've used the `uniqid()` function to generate the token name (https://github.com/AndrewCarterUK/Slim-Csrf/blob/master/src/Guard.php#L173)

I think it's probably worth reflecting both of those changes in this PR.

I would also add a unit test to check that the token has been generated as a hexadecimal string (`ctype_xdigit()`) that is the right length (`strlen()` should be double the $strength parameter I think).

**Comment by schnittstabil** [comment]
@AndrewCarterUK Thank you for your suggestions.

About 1.:  I would leave this up to the PO, the implementation is still based on the OWASP article and as long as we don't have a better article to reference I see no need to remove it.

About 2.: I think `uniqid()` does not have any benefits over `random_int`, in particular from [the `uniqid` docs](http://php.net/manual/en/function.uniqid.php):
> **Warning** This function does not create random nor unpredictable strings. This function must not be used for security purposes. …

<br>

You wrote:
> I would also add a unit test to check that the token has been generated as a hexadecimal string (ctype_xdigit()) that is the right length (strlen() should be double the $strength parameter I think).

I like the idea. As you already closed your PR, you may want to create a PR on my [Slim-Csrf-Fork](https://github.com/schnittstabil/Slim-Csrf) and we share the fame.

**Comment by AndrewCarterUK** [comment]
About 1: The recommendation to remove it was from @akrabat in IRC - but it might be worth pinging a few others on here to see their thoughts.

About 2: [`uniqid()`](http://php.net/manual/en/function.uniqid.php) is the right function to use because it guarantees that the token generated is unique (this is not done by `random_int()`. 

It doesn't matter if the name of the token is cryptographically secure (or predictable), as long as the value of the token is (otherwise what's the point of the pair?).

The requirement for the name of the token is uniqueness and the requirement for the value of the token is entropy (hence `uniqid()` and `random_bytes()`).

I'll see if I get time to submit the test tomorrow, it shouldn't take too long.

**Comment by schnittstabil** [comment]
@AndrewCarterUK You're right – I've switched to `uniqid`.

**Comment by AndrewCarterUK** [comment]
This looks good to me now :+1:

File: src/Guard.php [hunk 2]
```diff
@@ -10,10 +10,7 @@
 use Psr\Http\Message\ResponseInterface;
 
 /**
- * CSRF protection middleware based
- * on the OWASP example linked below.
- *
- * @link https://www.owasp.org/index.php/PHP_CSRF_Guard
+ * CSRF protection middleware.
```
Comment (schnittstabil) [inline src/Guard.php]: (Suggested by @akrabat and @AndrewCarterUK)

**Comment by schnittstabil** [comment]
Yeah, I think it is ready to merge.

**Comment by akrabat** [comment]
Thank you all for this work.

**Pull request merged**